src/app/modules/bounce/__tests__/bounce.service.spec.ts

```diff
@@ -106,499 +116,977 @@ describe('isValidSnsRequest', () => {
-  })
+  beforeAll(async () => await dbHandler.connect())
 
-  afterEach(async () => {
+  afterAll(async () => {
     await dbHandler.clearDatabase()
-    resetMockLogger(mockLogger)
-    resetMockLogger(mockShortTermLogger)
+    await dbHandler.closeDatabase()
   })
 
-  afterAll(async () => await dbHandler.closeDatabase())
-
-  it('should save correctly when there is a single delivery notification', async () => {
-    const formId = new ObjectId()
-    const submissionId = new ObjectId()
-    const notification = makeDeliveryNotification(
-      formId,
-      submissionId,
-      recipientList,
-      recipientList,
-    )
-    await updateBounces(notification)
-    const actualBounceDoc = await Bounce.findOne({ formId })
-    const actualBounce = extractBounceObject(actualBounceDoc!)
-    const expectedBounces = recipientList.map((email) => ({
-      email,
-      hasBounced: false,
-    }))
-    expect(mockLogger.info.mock.calls[0][0]).toMatchObject({
-      meta: {
-        ...JSON.parse(notification.Message),
-      },
+  describe('Non-critical bounces', () => {
+    beforeEach(() => {
+      jest.resetAllMocks()
     })
-    expect(mockLogger.warn).not.toHaveBeenCalled()
-    expect(omit(actualBounce, 'expireAt')).toEqual({
-      formId,
-      hasAlarmed: false,
-      bounces: expectedBounces,
-    })
-    expect(actualBounce.expireAt).toBeInstanceOf(Date)
-  })
 
-  it('should save correctly when there is a single non-critical bounce notification', async () => {
-    const bounces = {
-      [recipientList[0]]: true,
-      [recipientList[1]]: false,
-      [recipientList[2]]: false,
-    }
-    const formId = new ObjectId()
-    const submissionId = new ObjectId()
-    const notification = makeBounceNotification(
-      formId,
-      submissionId,
-      recipientList,
-      recipientList.slice(0, 1), // Only first email bounced
-    )
-    await updateBounces(notification)
-    const actualBounceDoc = await Bounce.findOne({ formId })
-    const actualBounce = extractBounceObject(actualBounceDoc!)
-    const expectedBounces = recipientList.map((email) => ({
-      email,
-      hasBounced: bounces[email],
-    }))
-    expect(mockLogger.info.mock.calls[0][0]).toMatchObject({
-      meta: {
-        ...JSON.parse(notification.Message),
-      },
-    })
-    expect(mockLogger.warn).not.toHaveBeenCalled()
-    expect(omit(actualBounce, 'expireAt')).toEqual({
-      formId,
-      hasAlarmed: false,
-      bounces: expectedBounces,
+    it('should save correctly when there is a single delivery notification', async () => {
+      const formId = new ObjectId()
+      const submissionId = new ObjectId()
+      const notification = makeDeliveryNotification(
+        formId,
+        submissionId,
+        MOCK_RECIPIENT_LIST,
+        MOCK_RECIPIENT_LIST,
+      )
+      await updateBounces(notification)
+      const actualBounceDoc = await Bounce.findOne({ formId })
+      const actualBounce = extractBounceObject(actualBounceDoc!)
```

Comment:
What happens if actualBounceDoc is in fact null?
```
export interface IBounceModel extends Model<IBounceSchema> {
  fromSnsNotification: (snsInfo: IEmailNotification) => IBounceSchema | null
}

```

Reply:
good point. this doesn't happen anymore since I moved the check for the email type into the controller.